README
A class for reading and writing *.wav files.

First, install the WaveFile gem...

    sudo gem install jstrait-wavefile -s http://gems.github.com

...and include it in your Ruby program:

    require 'wavefile'

To open a wave file and get the raw sample data:

    w = WaveFile.open("myfile.wav")
    samples = w.sample_data

You can also get the sample data in a normalized form, with each sample between -1.0 and 1.0:

    normalized_samples = w.normalized_sample_data

You can get basic metadata:

    w.num_channels      # 1 for mono, 2 for stereo
    w.sample_rate       # 11025, 22050, 44100, etc.
    w.bits_per_sample   # 8 or 16

To create and save a new wave file:

    w = WaveFile.new(1, 44100, 16)  # num_channels,
                                    # sample_rate,
                                    # bits_per_sample
    w.sample_data = <array of samples goes here>
    w.save("myfile.wav")

When calling the sample_data=() method, the passed in array can contain either raw samples or normalized samples. If the first item in the array is a Float, the entire array is assumed to be normalized. Normalized samples are automatically converted into raw samples when saving.


CHANGELOG
1.1.2
===============================================================================
* **Bug Fix:** Files that have extra bytes at the end of the `"fmt "` chunk can
  now be read. If the format code is `1`, the `"fmt "` chunk has extra bytes if
  the chunk body size is greater than 16 bytes. Otherwise, "extra bytes" means
  the chunk contains bytes after the chunk extension (not including the
  required padding byte for an odd-sized chunk). Previously, attempting to open
  certain files like this via `Reader.new` would result in `InvalidFormatError`
  being raised with a misleading `"Not a supported wave file. The format chunk
  extension is shorter than expected."` message. This was misleading because if
  the format code is `1`, the `"fmt "` chunk won't actually have a chunk
  extension, and for other format codes the chunk extension might actually be
  the expected size or larger. When reading a file like this, any extra data in
  the `"fmt "` chunk beyond what is expected based on the relevant format code
  will now be ignored.
  * There was a special case where a file like this _could_ be opened
    correctly. If the format code was `1`, and the value of bytes 16 and 17
    (0-based), when interpreted as a 16-bit unsigned little-endian integer,
    happened to be the same as the number of subsequent bytes in the chunk, the
    file could be opened without issue. For example, if the `"fmt "` chunk size
    was `22`, the format code was `1`, and the value of bytes 16 and 17 was `4`
    (when interpreted as a 16-bit unsigned little-endian integer), the file
    could be opened correctly.
  * There was another special case where `InvalidFormatError` would be
    incorrectly raised, but the error message would be different (and also
    misleading). If the format code was `1`, and there was exactly 1 extra byte
    in the `"fmt "` chunk (i.e. the chunk size was 17 bytes), the error message
    would be `"Not a supported wave file. The format chunk is missing an
    expected extension."` This was misleading because when the format code is
    `1`, the `"fmt "` chunk doesn't have a chunk extension.
  * Thanks to [@CromonMS](https://github.com/CromonMS) for reporting this as an
    issue.
* **Bug Fix:** Files in WAVE_FORMAT_EXTENSIBLE format with a missing or
  incomplete `"fmt "` chunk extension can no longer be opened using
  `Reader.new`. Previously, a `Reader` instance could be constructed for a file
  like this, but the relevant fields on the object returned by
  `Reader#native_format` would contain `nil` or `""` values for these fields,
  and no sample data could be read from the file. Since files like this are
  missing required fields that don't necessarily have sensible default values,
  it seems like it shouldn't be possible to create a `Reader` instance from
  them. After this fix, attempting to do so will cause `InvalidFormatError` to
  be raised.
* **Bug Fix:** Files in WAVE_FORMAT_EXTENSIBLE format that have extra bytes at
  the end of the `"fmt "` chunk extension can now be read. This is similar but
  different from the first bug above; that bug refers to extra bytes _after_
  the chunk extension, while this bug refers to extra bytes _inside_ the chunk
  extension. A WAVE_FORMAT_EXTENSIBLE `"fmt "` chunk extension has extra bytes
  if it is larger than 22 bytes. Previously, a `Reader` instance could be
  constructed for a file like this, but
  `Reader#native_format#sub_audio_format_guid` would have an incorrect value,
  and sample data could not be read from the file. After this fix, this field
  will have the correct value, and if it is one of the supported values then
  sample data can be read. Any extra data at the end of the chunk extension
  will be ignored. Implicit in this scenario is that the `"fmt "` chunk has a
  stated size large enough to fit the oversized chunk extension. For cases
  where it doesn't, see the next bug fix below.
* **Bug Fix:** More accurate message on the `InvalidFormatError` raised when
  reading a file whose `"fmt "` chunk extension is too large to fit in the
  chunk. The message will now correctly state that the chunk extension is too
  large, rather than `"Not a supported wave file. The format chunk extension is
  shorter than expected."`. As an example of what "too large" means, if a
  `"fmt "` chunk has a size of 50 bytes, then any chunk extension larger than
  32 bytes will be too large and overflow out of the chunk, since a chunk
  extension's content always starts at byte 18 (0-based).


1.1.1
===============================================================================
* Removes `warning: Using the last argument as keyword parameters is
  deprecated; maybe ** should be added to the call` output when reading a
  file with a `smpl` chunk using Ruby 2.7.0. (And presumably, higher Ruby
  versions as well, but Ruby 2.7.0 is the most recent Ruby version at the
  time of this release).


1.1.0
===============================================================================
* **Can read `smpl` chunk data from files that contain this kind of chunk.**
  If a *.wav file contains a `smpl` chunk, then `Reader.sampler_info` will
  return a `SamplerInfo` instance with the relevant data (or `nil` otherwise).
  Thanks to [@henrikj242](https://github.com/henrikj242) for suggesting this
  feature and providing the base implementation.
* **More informative errors raised by `Reader.new`**. When attempting to read
  an invalid file, the error message now provides more detail about why the
  file is invalid.
* **Bug Fix**: The master RIFF chunk size for files written by the gem will
  now take into account padding bytes written for child chunks. For example,
  when writing a file with a `data` chunk whose body has an odd number of
  bytes, the master RIFF chunk's size will be 1 byte larger (to take the empty
  padding byte at the end of the `data` chunk into account).
* **Bug Fix**: If the stated `data` chunk size is larger than the actual number

tools/debug.rb
CHUNK_BODY_READERS = {
  "RIFF" => :read_riff_chunk,
  "fmt " => :read_format_chunk,
  "fact" => :read_fact_chunk,
  "PEAK" => :read_peak_chunk,
  "cue " => :read_cue_chunk,
  "smpl" => :read_sample_chunk,
  "inst" => :read_instrument_chunk,
  "LIST" => :read_list_chunk,
  "data" => :read_data_chunk,
}
CHUNK_BODY_READERS.default = :read_unrecognized_chunk

def main
  File.open(ARGV[0], "rb") do |file|
    field_reader = FieldReader.new(file)

    begin
      begin
        riff_chunk_id_field = field_reader.read_fourcc
        riff_chunk_size_field = field_reader.read_uint32
      ensure
        if riff_chunk_id_field != nil
          display_chunk_header(riff_chunk_id_field, riff_chunk_size_field)
        end
      end

      read_riff_chunk(field_reader, riff_chunk_size_field[:parsed_value])
    rescue EOFError
      # Swallow the error and do nothing to avoid an error being shown in the output.
      # Perhaps in the future it would be better to show an indication that the end
      # of the file was unexpectedly reached.
    end
  end
end


def display_line(label, field)
  parsed_value = field[:parsed_value]
  bytes = field[:bytes]
  data_type = field[:type_label]

  if data_type == "FourCC" || data_type == "C String"
    # Wrap the value in quotes and show character codes for non-display characters
    formatted_value = parsed_value.inspect
  else
    # This branch exists to avoid wrapping a value in quotes when it semantically
    # is not a String but happens to be contained in a String object (e.g. a bit field,
    # GUID, etc).
    formatted_value = parsed_value.to_s
  end

  formatted_bytes = bytes.map {|byte| byte.unpack("H2").first }.join(" ")

  puts "#{(label + ":").ljust(22)} #{data_type.ljust(10)} | #{formatted_value.ljust(10)} | #{formatted_bytes}"
end


def display_chunk_header(chunk_id_field, chunk_size_field)
  title = "#{chunk_id_field[:parsed_value].inspect} Chunk"

  if chunk_id_field[:parsed_value] == "RIFF"
    title += " Header"
  end

  puts title
  puts "=================================================================================="
  display_line("Chunk ID", chunk_id_field)

  if chunk_size_field != nil
    display_line("Chunk size", chunk_size_field)
    display_chunk_section_separator
  end
end


def display_chunk_section_separator
  puts "----------------------------------+------------+----------------------------------"
end


def read_riff_chunk(field_reader, chunk_size)
  display_line("Form type", field_reader.read_fourcc)

  while field_reader.bytes_read < chunk_size + 8
    child_chunk_id_field, child_chunk_size_field = nil

    begin
      child_chunk_id_field = field_reader.read_fourcc
      child_chunk_size_field = field_reader.read_uint32
    ensure
      if child_chunk_id_field != nil
        puts ""
        puts ""

        display_chunk_header(child_chunk_id_field, child_chunk_size_field)
      end
    end

    chunk_body_reader_method_name = CHUNK_BODY_READERS[child_chunk_id_field[:parsed_value]]
    send(chunk_body_reader_method_name, field_reader, child_chunk_size_field[:parsed_value])

    if child_chunk_size_field[:parsed_value].odd?
      display_line("Padding Byte", field_reader.read_padding_byte)
    end
  end
end


def read_format_chunk(field_reader, chunk_size)
  audio_format_code_field = field_reader.read_uint16
  display_line("Audio format", audio_format_code_field)
  display_line("Channels", field_reader.read_uint16)
  display_line("Sample rate", field_reader.read_uint32)
  display_line("Byte rate", field_reader.read_uint32)
  display_line("Block align", field_reader.read_uint16)
  display_line("Bits per sample", field_reader.read_uint16)

  bytes_read_so_far = 16

  if audio_format_code_field[:parsed_value] != 1 && chunk_size > 16
    extension_size_field = field_reader.read_uint16
    display_line("Extension size", extension_size_field)
    bytes_read_so_far += 2

    if extension_size_field[:parsed_value] > 0
      if audio_format_code_field[:parsed_value] == 65534
        display_line("Valid bits per sample", field_reader.read_uint16)
        display_line("Speaker mapping", field_reader.read_bitfield(4))
        display_line("Sub format GUID", field_reader.read_guid)

        extra_byte_count = extension_size_field[:parsed_value] - 22
        if extra_byte_count > 0
          display_line("Extra extension bytes", field_reader.read_bytes(extra_byte_count))
        end
      else
        display_line("Raw extension", field_reader.read_bytes(extension_size_field[:parsed_value]))
      end
    end

    bytes_read_so_far += extension_size_field[:parsed_value]
  end

  extra_byte_count = chunk_size - bytes_read_so_far
  if extra_byte_count > 0
    display_line("Extra bytes", field_reader.read_bytes(extra_byte_count))
  end
end


def read_fact_chunk(field_reader, chunk_size)
  display_line("Sample count", field_reader.read_uint32)

  if chunk_size > 4
    display_line("Extra bytes", field_reader.read_bytes(chunk_size - 4))
  end
end


def read_peak_chunk(field_reader, chunk_size)
  display_line("Version", field_reader.read_uint32)
  display_line("Timestamp", field_reader.read_uint32)

  ((chunk_size - 8) / 8).times do |i|
    display_line("Chan. #{i + 1} Value", field_reader.read_float32)
    display_line("Chan. #{i + 1} Position", field_reader.read_uint32)
  end
end


def read_cue_chunk(field_reader, chunk_size)
  cue_point_count_field = field_reader.read_uint32
  display_line("Cue point count", cue_point_count_field)

  bytes_remaining = chunk_size - 4

  cue_point_count_field[:parsed_value].times do |i|
    display_line("ID #{i + 1}", field_reader.read_uint32)
    display_line("Position #{i + 1}", field_reader.read_uint32)
    display_line("Chunk type #{i + 1}", field_reader.read_fourcc)
    display_line("Chunk start #{i + 1}", field_reader.read_uint32)
    display_line("Block start #{i + 1}", field_reader.read_uint32)
    display_line("Sample offset #{i + 1}", field_reader.read_uint32)
    bytes_remaining -= 24
  end

  if bytes_remaining > 0
    display_line("Extra bytes", field_reader.read_bytes(bytes_remaining))
  end
end


def read_sample_chunk(field_reader, chunk_size)
  display_line("Manufacturer", field_reader.read_uint32)
  display_line("Product", field_reader.read_uint32)
  display_line("Sample Period", field_reader.read_uint32)
  display_line("MIDI Unity Note", field_reader.read_uint32)
  display_line("MIDI Pitch Fraction", field_reader.read_uint32)
  display_line("SMPTEFormat", field_reader.read_uint32)
  display_line("SMPTEOffset", field_reader.read_uint32)

  loop_count_field = field_reader.read_uint32
  loop_count = loop_count_field[:parsed_value]
  display_line("Sample Loops", loop_count_field)

  sampler_specific_data_size_field = field_reader.read_uint32
  sampler_specific_data_size = sampler_specific_data_size_field[:parsed_value]
  display_line("Sampler Data Size", sampler_specific_data_size_field)

  loop_count.times do |i|
    display_chunk_section_separator
    puts "Loop ##{i + 1}:"
    display_line("Identifier", field_reader.read_uint32)
    display_line("Type", field_reader.read_uint32)
    display_line("Start", field_reader.read_uint32)
    display_line("End", field_reader.read_uint32)
    display_line("Fraction", field_reader.read_uint32)
    display_line("Play Count", field_reader.read_uint32)
  end

  if sampler_specific_data_size > 0
    display_chunk_section_separator
    display_line("Sampler specific data", field_reader.read_bytes(sampler_specific_data_size))
  end

  extra_byte_count = chunk_size - 36 - (loop_count * 24) - sampler_specific_data_size_field[:parsed_value]
  if (extra_byte_count > 0)
    display_line("Extra bytes", field_reader.read_bytes(extra_byte_count))
  end
end


def read_instrument_chunk(field_reader, chunk_size)
  display_line("Unshifted Note", field_reader.read_uint8)
  display_line("Fine Tune", field_reader.read_int8)
  display_line("Gain", field_reader.read_int8)
  display_line("Low Note", field_reader.read_uint8)
  display_line("High Note", field_reader.read_uint8)
  display_line("Low Velocity", field_reader.read_uint8)
  display_line("High Velocity", field_reader.read_uint8)

  extra_data_size = chunk_size - 7
  if extra_data_size > 0
    display_line("Extra bytes", field_reader.read_bytes(extra_data_size))
  end
end


def read_list_chunk(field_reader, chunk_size)
  list_type_field = field_reader.read_fourcc
  display_line("List Type", list_type_field)

  bytes_remaining = chunk_size - 4

  while bytes_remaining > 0
    display_chunk_section_separator
    child_chunk_id_field = field_reader.read_fourcc
    display_line("Child Chunk ID", child_chunk_id_field)

    child_chunk_size_field = field_reader.read_uint32
    child_chunk_size = child_chunk_size_field[:parsed_value]

    display_line("Child Chunk Size", child_chunk_size_field)

    if list_type_field[:parsed_value] == "adtl"
      display_line("Cue Point ID", field_reader.read_uint32)

      if child_chunk_id_field[:parsed_value] == "ltxt"
        display_line("Sample Length", field_reader.read_uint32)
        display_line("Purpose", field_reader.read_fourcc)
        display_line("Country", field_reader.read_uint16)
        display_line("Language", field_reader.read_uint16)
        display_line("Dialect", field_reader.read_uint16)
        display_line("Code Page", field_reader.read_uint16)

        if child_chunk_size > 20
          display_line("Text", field_reader.read_bytes(child_chunk_size - 20))
        end
      else
        display_line("Content", field_reader.read_null_terminated_string(child_chunk_size - 4))
      end

      bytes_remaining -= (child_chunk_size + 8)
    else   # INFO, and any unknown list type
      display_line("Content", field_reader.read_null_terminated_string(child_chunk_size))

      bytes_remaining -= (child_chunk_size + 8)
    end

    if child_chunk_size.odd?
      display_line("Padding Byte", field_reader.read_padding_byte)
      bytes_remaining -= 1
    end
  end
end


def read_data_chunk(field_reader, chunk_size)
  intro_byte_count = [10, chunk_size].min

  if intro_byte_count > 0
    display_line("Data Start", field_reader.read_bytes(intro_byte_count))
  end

  if intro_byte_count < chunk_size
    skipped_byte_count = field_reader.skip_bytes(chunk_size - intro_byte_count)
    puts "(#{skipped_byte_count} bytes omitted)"
  end
end


def read_unrecognized_chunk(field_reader, chunk_size)
  if chunk_size > 0
    skipped_byte_count = field_reader.skip_bytes(chunk_size)
    puts "(unrecognized chunk type; #{skipped_byte_count} byte chunk body omitted)"
  end
end


class FieldReader
  def initialize(file)
    @file = file
    @bytes_read = 0
  end

  attr_reader :bytes_read

  def read_int8
    read_field(byte_count: 1,
               type_label: "int_8",
               parser: lambda {|bytes| bytes.join.unpack("c").first })
  end

  def read_uint8
    read_field(byte_count: 1,
               type_label: "uint_8",
               parser: lambda {|bytes| bytes.join.unpack("C").first })
  end

  def read_uint16
    read_field(byte_count: 2,
               type_label: "uint_16",
               parser: lambda {|bytes| bytes.join.unpack("v").first })
  end

  def read_uint32
    read_field(byte_count: 4,
               type_label: "uint_32",
               parser: lambda {|bytes| bytes.join.unpack("V").first })
  end

  def read_float32
    read_field(byte_count: 4,
               type_label: "float_32",
               parser: lambda {|bytes| bytes.join.unpack("e").first })
  end

  def read_fourcc
    read_field(byte_count: 4,
               type_label: "FourCC",
               parser: lambda {|bytes| bytes.join })
  end

  def read_null_terminated_string(byte_count)
    read_field(byte_count: byte_count,
               type_label: "C String",
               parser: lambda {|bytes| bytes.join.unpack("Z#{byte_count}").first })
  end

  def read_bitfield(byte_count)
    read_field(byte_count: byte_count,
               type_label: "Bit field",
               parser: lambda {|bytes| bytes.reverse.map {|byte| byte.unpack("B8")}.join })
  end

  def read_guid
    read_field(byte_count: 16,
               type_label: "GUID",
               parser: lambda {|bytes| "0x#{bytes.map {|byte| byte.unpack("H2")}.join}" })
  end

  def read_bytes(byte_count)
    read_field(byte_count: byte_count,
               type_label: "bytes",
               parser: lambda {|bytes| "N/A" })
  end

  def read_padding_byte
    read_field(byte_count: 1,
               type_label: "byte",
               parser: lambda {|bytes| bytes.join.unpack("C").first })
  end

  def skip_bytes(byte_count)
    string = @file.sysread(byte_count)

    @bytes_read += string.length
    string.length
  end

  private

  def read_field(byte_count: nil, type_label: nil, parser: nil)
    bytes = read_field_bytes(byte_count)

    {
      bytes: bytes,
      type_label: type_label,
      parsed_value: parser.call(bytes)
    }
  end

  def read_field_bytes(byte_count)
    bytes = []

    byte_count.times do |i|
      bytes << @file.sysread(1)
      @bytes_read += 1
    end

    bytes
  end
end


main
